### PR TITLE
New component: Counter #785

### PR DESCRIPTION
**Pull request by aripddev**
Completed

### CODE REVIEW

File: src/main/java/org/primefaces/extensions/component/counter/Counter.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.primefaces.extensions.component.counter;
+
+import javax.faces.application.ResourceDependency;
+import javax.faces.component.FacesComponent;
+
+@FacesComponent(value = Counter.COMPONENT_TYPE)
+@ResourceDependency(library = "primefaces", name = "components.css")
```
Comment (melloware) [inline src/main/java/org/primefaces/extensions/component/counter/Counter.java:30]: Remove @FacesComponent I found this caused a problem with FuzzySearch

File: src/main/java/org/primefaces/extensions/component/counter/CounterBase.java [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+import javax.faces.component.UIComponentBase;
+import org.primefaces.component.api.Widget;
+
+public abstract class CounterBase extends UIComponentBase implements Widget {
+
+    public static final String COMPONENT_FAMILY = "org.primefaces.extensions.component";
+
+    public static final String DEFAULT_RENDERER = "org.primefaces.extensions.component.CounterRenderer";
+
```
Comment (melloware) [inline src/main/java/org/primefaces/extensions/component/counter/CounterBase.java:28]: Add `start` and `end` EVENTS with the default event being `end`

File: src/main/java/org/primefaces/extensions/component/counter/CounterHandler.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+import javax.faces.view.facelets.ComponentHandler;
+import javax.faces.view.facelets.MetaRule;
+import javax.faces.view.facelets.MetaRuleset;
+
+import org.primefaces.facelets.MethodRule;
+
+/**
+ * Handler for {@link Counter}
+ */
+public class CounterHandler extends ComponentHandler {
```
Comment (melloware) [inline src/main/java/org/primefaces/extensions/component/counter/CounterHandler.java]: What does this class do?

File: src/main/java/org/primefaces/extensions/component/counter/CounterRenderer.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+import javax.faces.context.ResponseWriter;
+import javax.faces.render.FacesRenderer;
+import org.primefaces.extensions.util.Attrs;
+import org.primefaces.renderkit.CoreRenderer;
+import org.primefaces.util.Constants;
+import org.primefaces.util.LangUtils;
+import org.primefaces.util.WidgetBuilder;
+
+@FacesRenderer(componentFamily = Counter.COMPONENT_FAMILY, rendererType = Counter.DEFAULT_RENDERER)
+public class CounterRenderer extends CoreRenderer {
```
Comment (melloware) [inline src/main/java/org/primefaces/extensions/component/counter/CounterRenderer.java:28]: Remove `@FacesRenderer` the declaration is the XML is all we need.

File: src/main/java/org/primefaces/extensions/component/counter/CounterRenderer.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    @Override
+    public void encodeChildren(FacesContext context, UIComponent component) throws IOException {
+        //Do nothing
+    }
+
+    @Override
+    public boolean getRendersChildren() {
+        return true;
+    }
+
```
Comment (melloware) [inline src/main/java/org/primefaces/extensions/component/counter/CounterRenderer.java]: I think you can remove the above two overrides?

File: src/main/resources/META-INF/resources/primefaces-extensions/counter/1-counter-widget.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+        if (this.autoStart) {
+            this.start();
+        }
+
+    },
+
+    start: function () {
+        var that = this;
+
```
Comment (melloware) [inline src/main/resources/META-INF/resources/primefaces-extensions/counter/1-counter-widget.js:46]: PrimeFaces uses `var $this` as standard not `that`

File: src/main/resources/META-INF/resources/primefaces-extensions/counter/1-counter-widget.js [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        if (!that.counter.error) {
+            if (that.oncountercomplete) {
+                that.counter.start(that.oncountercomplete);
+            } else {
+                that.counter.start();
+            }
```
Comment (melloware) [inline src/main/resources/META-INF/resources/primefaces-extensions/counter/1-counter-widget.js:52]: `this.callBehavior('start');`

File: src/main/resources/META-INF/resources/primefaces-extensions/counter/1-counter-widget.js [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+     */
+    init: function (cfg) {
+
+        this._super(cfg);
+        this.id = cfg.id;
+        this.cfg = cfg;
+
+        this.end = cfg.end;
+        this.autoStart = cfg.autoStart;
+        this.oncountercomplete = cfg.oncountercomplete;
```
Comment (melloware) [inline src/main/resources/META-INF/resources/primefaces-extensions/counter/1-counter-widget.js:21]: Need to wrap this oncountercomplete so you can call `this.callBehavior('end');` whether the user passes in a custom function or not

**Review by melloware — CHANGES_REQUESTED** [review]

File: src/main/java/org/primefaces/extensions/component/counter/CounterHandler.java (hunk 3 above)
Comment (aripdcem) [reply src/main/java/org/primefaces/extensions/component/counter/CounterHandler.java]: I tried to add listener, then forgot to delete it

File: src/main/resources/META-INF/resources/primefaces-extensions/counter/1-counter-widget.js (hunk 8 above)
Comment (aripdcem) [reply src/main/resources/META-INF/resources/primefaces-extensions/counter/1-counter-widget.js:21]: I did not understand this :(

Comment (melloware) [reply src/main/resources/META-INF/resources/primefaces-extensions/counter/1-counter-widget.js:21]: I can take care of this one don't worry about it.

**Comment by melloware** [comment]
Nice work on this.  You satisfied enough for me to merge it and test it out?

**Comment by aripddev** [comment]
Thanks @melloware, you can merge it if it is appropriate.

**Pull request merged**